x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_hosts_url_state.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useReducer } from 'react';
+import { TimeRange } from '@kbn/es-query';
+import DateMath from '@kbn/datemath';
+import deepEqual from 'fast-deep-equal';
+import * as rt from 'io-ts';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { fold } from 'fp-ts/lib/Either';
+import { constant, identity } from 'fp-ts/lib/function';
+import { enumeration } from '@kbn/securitysolution-io-ts-types';
+import { FilterStateStore } from '@kbn/es-query';
+import { useUrlState } from '../../../../utils/use_url_state';
+import { useKibanaTimefilterTime } from '../../../../hooks/use_kibana_timefilter_time';
+
+const DEFAULT_FROM_MINUTES_VALUE = 15;
+
+const INITIAL_DATE = new Date();
+export const INITIAL_DATE_RANGE = { from: 'now-15m', to: 'now' };
+export const DEFAULT_QUERY = {
+  language: 'kuery',
+  query: '',
+};
+
+// with this approach, all state will be in one variable in the url
+const INITIAL_HOSTS_STATE: HostsState = {
+  query: DEFAULT_QUERY,
+  filters: [],
+  // for unified search
+  dateRange: { ...INITIAL_DATE_RANGE },
+  // for useSnapshot
+  dateRangeTimestamp: {
+    from: new Date(INITIAL_DATE.getMinutes() - DEFAULT_FROM_MINUTES_VALUE).getTime(),
+    to: INITIAL_DATE.getTime(),
+  },
+};
+
+type Action =
+  | {
+      type: 'setQuery';
+      payload: rt.TypeOf<typeof SetQueryType>;
+    }
+  | { type: 'setFilter'; payload: rt.TypeOf<typeof HostsFiltersRT> };
+
+const reducer = (state: HostsState, action: Action): HostsState => {
+  switch (action.type) {
+    case 'setFilter':
+      return { ...state, filters: [...action.payload] };
+    case 'setQuery':
+      const { filters, query, ...payload } = action.payload;
+      const newFilters = !filters ? state.filters : filters;
+      const newQuery = !query ? state.query : query;
+      return {
+        ...state,
+        ...payload,
+        filters: [...newFilters],
+        query: { ...newQuery },
+      };
+    default:
+      throw new Error();
+  }
+};
+
+export const useHostsUrlState = () => {
+  const [urlState, setUrlState] = useUrlState<HostsState>({
+    defaultState: INITIAL_HOSTS_STATE,
+    decodeUrlState,
+    encodeUrlState,
+    urlStateKey: '_a', // TBD
+  });
+
+  const [state, dispatch] = useReducer(reducer, urlState);
+
+  const [getTime] = useKibanaTimefilterTime(INITIAL_DATE_RANGE);
+
+  const getRangeInTimestamp = useCallback(({ from, to }: TimeRange) => {
+    const fromTS = DateMath.parse(from)?.valueOf() ?? 0;
+    const toTS = DateMath.parse(to)?.valueOf() ?? 0;
```

Comment:
This is probably an edge case, but in the unlikely event of `form` or `to` are `undefined`, we might want to calculate a fallback date range using `new Date` - similar to https://github.com/elastic/kibana/pull/144181/files#diff-46e56847f0a8bf5427cc4c1c93d280065ee9256862d4f553179dec3e8ddd7d61R37, instead of returning `0`

Reply:
Good idea! I moved the initial calculation to a constant and reused it here and inside `dateRangeTimestamp`.